cli/src/maintenance.rs

```diff
@@ -571,32 +614,53 @@ impl SolidoState {
     /// or if some joker donates to one of the stake accounts we can use the same function
     /// to claim these rewards back to the reserve account so they can be re-staked.
     pub fn try_withdraw_inactive_stake(&self) -> Option<(Instruction, MaintenanceOutput)> {
-        for (validator, stake_accounts) in self
-            .solido
-            .validators
-            .entries
-            .iter()
-            .zip(self.validator_stake_accounts.iter())
-        {
-            let current_balance = stake_accounts
+        for (validator, stake_accounts, unstake_accounts) in izip!(
+            self.solido.validators.entries.iter(),
+            self.validator_stake_accounts.iter(),
+            self.validator_unstake_accounts.iter()
+        ) {
+            let current_stake_balance = stake_accounts
                 .iter()
                 .map(|(_addr, detail)| detail.balance.total())
                 .sum::<token::Result<Lamports>>()
                 .expect("If this overflows, there would be more than u64::MAX staked.");
 
-            if current_balance > validator.entry.stake_accounts_balance {
-                let expected_difference = (current_balance
-                    - validator.entry.stake_accounts_balance)
-                    .expect("Does not overflow because current > entry.balance.");
-                // If the expected difference is less than some defined amount
-                // of Lamports, we don't bother withdrawing. We try to do this
-                // so we don't pay more for fees than the amount that we'll
-                // withdraw.
-                if expected_difference < SolidoState::MINIMUM_WITHDRAW_AMOUNT {
-                    continue;
+            let expected_difference_stake =
```

Comment:
This looks pretty complicated, and now that I’m looking at the original code, I’m not sure if it was doing the right thing ... it checked for donations but withdrawing inactive stake can still be worthwhile even if the total balance isn’t the same ...

There are two cases in which we want to call `WithdrawInactiveStake`:

* If the stake accounts balance stored doesn’t match the balance observed, then `WithdrawInactiveStake` will fix it.
* If there is inactive stake that we can withdraw, then we should.

So how about this:

* Loop over all stake accounts and unstake accounts, and compute the total balances, and the amount of withdrawable inactive stake (`inactive - rent` for stake accounts, `inactive` for unstake accounts).
* If the computed total does not match, or if the withdrawable amount is greater than the minimum withdraw amount, continue to call the intruction.